Add endpoint to get initiatives by location with coordinates

denarch22 opened the pull request:
## 🛠️ Changes
Introduces a new DTO (InitiativeCoordinatesDto), repository method, service method, and API endpoint to retrieve active initiatives with their coordinates, optionally filtered by location. Also adds a Swagger example response for documentation.

## 📝 Associated issues

Resolve [LIB-352](https://linear.app/biodev/issue/LIB-352/monitoreo-comunitario-endpoint-de-coordenadas-de-iniciativas)

## ✅ Checks
Update examples

denarch22 committed:
Add endpoint to get initiatives by location with coordinates

Introduces a new DTO (InitiativeCoordinatesDto), repository method, service method, and API endpoint to retrieve active initiatives with their coordinates, optionally filtered by location. Also adds a Swagger example response for documentation.

denarch22 committed:
Remove exception message from initiative location logging

Updated InitiativeService to exclude exception messages from logs when errors occur during location-based initiative retrieval. Minor formatting fixes applied to InitiativeRepository and InitiativeLocationResponseExample for consistency.

denarch22 committed:
Fix BOM and typo in Initiative files

Removed BOM from InitiativeLocationResponseExample.cs and added BOM to InitiativeRepository.cs for consistency. Also corrected a typo in the initiative name from 'Conservación Amazonas' to 'Conservacion Amazonas'.

denarch22 committed:
Refactor GetExamples to use collection expressions

Updated the GetExamples method to use C# collection expressions for list and array initialization, improving code readability and conciseness.

cgalvist commented on src/Application/Services/Initiatives/InitiativeService.cs:
```diff
@@ -473,6 +473,39 @@ public async Task<CustomWebResponse> UpdatePolygonAsync(int id, string geoJsonSt
+                ResponseBody = result,
+            };
+        }
+        catch (Exception)
```
¿En qué casos puede haber una excepción? ¿y de qué tipo?
Porfa, creo que sería mejor ser más específicos con as excepciones que se puedan presentar

denarch22 replied:
Gracias por la observación sobre el manejo de excepciones. Tienes razón, es importante ser más específicos con los tipos de excepciones.

He implementado una solución más robusta y escalable:
Solución Implementada: Middleware Global de Manejo de Excepciones
En lugar de manejar excepciones específicas en cada método del servicio, he creado un ExceptionHandlingMiddleware que centraliza el manejo de errores en toda la aplicación.
Beneficios de esta implementación:
Manejo específico por tipo de excepción:
OperationCanceledException → HTTP 408 (Request Timeout)
InvalidOperationException → HTTP 400 (Bad Request)
ArgumentException → HTTP 400 (Bad Request)
TimeoutException → HTTP 408 (Request Timeout)
DbUpdateException → HTTP 500 (Internal Server Error)
UnauthorizedAccessException → HTTP 401 (Unauthorized)
NotImplementedException → HTTP 501 (Not Implemented)
Logging estructurado: Cada excepción se registra con el nivel apropiado y información detallada
Consistencia: Todas las excepciones se manejan de manera uniforme en toda la aplicación
Mantenibilidad: Un solo lugar para actualizar el manejo de errores
Código más limpio: Los servicios se enfocan en la lógica de negocio
Archivos modificados:
✅ src/WebApi/Middleware/ExceptionHandlingMiddleware.cs (nuevo)
✅ src/WebApi/Program.cs (registro del middleware)
✅ src/Application/Services/Initiatives/InitiativeService.cs (simplificado)
Esta solución responde directamente a tu comentario sobre la necesidad de ser más específicos con las excepciones, pero de una manera más elegante y mantenible que beneficia a toda la aplicación.

cgalvist reviewed (changes requested):
Hola.
También te pido cambiar la rama destino a `darmenta/lib-351`, para poder revisar únicamente los cambios de este ticket.

denarch22 committed:
Add global exception handling middleware

Introduced ExceptionHandlingMiddleware for centralized error handling in the Web API. Updated Program.cs to register the middleware in the request pipeline. Refactored InitiativeService.GetByLocationAsync to remove internal try-catch, delegating exception handling to the middleware.

denarch22 committed:
Refactor statistics service to unify logic

Consolidated general statistics retrieval into a single internal method with optional filters for department and initiative. Removed redundant private methods and simplified public API, improving maintainability and reducing code duplication.

denarch22 committed:
Update GeneralStatisticsController.cs

cgalvist commented on src/WebApi/Config/DocsSetup/Examples/Initiative/InitiativeLocationResponseExample.cs:
```diff
@@ -0,0 +1,38 @@
+/// <summary>
+/// Initiative location response example.
+/// </summary>
+public class InitiativeLocationResponseExample : IExamplesProvider<List<InitiativeCoordinatesDto>>
```
Para evitar confusiones con la entidad `InitiativeLocation`, recomiendo cambiar el nombre de esta clase a `InitiativeGeoDataResponseExample` u otro nombre diferente

cgalvist commented on src/Application/DTOs/Initiatives/InitiativeCoordinatesDto.cs:
```diff
@@ -0,0 +1,26 @@
+/// <summary>
+/// Initiative coordinates data transfer object.
+/// </summary>
+public class InitiativeCoordinatesDto : IDto
```
Sé que esta clase es un DTO, pero para poder utilizarla en la capa de infraestructura, recomiendo cambiarle el nombre a `InitiativeGeoData` y dejarla en `src/Core/Domain/Entities/Initiatives/InitiativeGeoData.cs`.
Luego de esto, cambiar los tipos anónimos de la capa de infraestructura para que usen esta clase.

cgalvist replied:
Veo que creaste la clase, pero `InitiativeCoordinatesDto` e `InitiativeGeoData` son clases prácticamente idénticas.

Ya que son representaciones de datos controladas y no son entidades, no veo necesario tener DTO por el momento.

¿Qué opinas?

cgalvist commented on src/Infrastructure/Persistence/Repositories/Initiatives/InitiativeRepository.cs:
```diff
@@ -152,4 +153,36 @@ public async Task<int> GetPeopleInvolvedInActiveInitiativesCountByInitiativeAsyn
+        return results.Select(r => (
+            r.Id,
+            r.Name,
+            r.Coordinate != null ? new double[] { r.Coordinate.X, r.Coordinate.Y } : System.Array.Empty<double>()));
```
ya que en la línea 166 se está validando que solo se traigan registros con coordenadas, creo que esto se puede simplificar con el código `new double[] { r.Coordinate.X, r.Coordinate.Y }`

cgalvist commented on src/WebApi/Middleware/ExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,131 @@
+    /// </summary>
+    /// <param name="context">HTTP context.</param>
+    /// <returns>Task.</returns>
+    public async Task InvokeAsync(HttpContext context)
```
Esta función no se está utilizando

cgalvist commented on src/WebApi/Middleware/ExceptionHandlingMiddleware.cs:
```diff
@@ -0,0 +1,131 @@
+    /// <param name="context">HTTP context.</param>
+    /// <param name="exception">Exception to handle.</param>
+    /// <returns>Task.</returns>
+    private async Task HandleExceptionAsync(HttpContext context, Exception exception)
```
Me parece que usar un Middleware para manejar los errores puede estar bien para muchos casos, pero hay que tener en cuenta que hay muchos tipos de error que estamos manejando de manera manual o que los manejan otros middlewares (por ejemplo, `Unauthorized`, `Forbidden`)

También el manejo de la excepción `UnauthorizedAccessException` puede generar un mensaje erroneo, ya que por ejemplo, si un usuario consume un endpoint que accede a un archivo en el backend, y el backend no tiene permisos para leer ese archivo, dispararía esa excepción pero retornaría un error `Unauthorized` con código 401, lo que para el frontend significa que el token de autorización es inválido o ya caducó.

Como dije antes, me parece buena idea usar un middleware para algunas excepciones que no estemos controlando, pero es algo que debemos probar con cuidado y siguiendo algunas normas. Si quieres, haz un nuevo ticket para añadir esta funcionalidad y probarla más adelante, pero no recomiendo agregar el midleware en este PR.

denarch22 committed:
Refactor initiative geo data handling and cleanup middleware

Introduced InitiativeGeoData entity and updated repository and service methods to use it for returning initiative coordinates. Renamed related API documentation example classes. Removed ExceptionHandlingMiddleware and its registration from the WebApi project.

denarch22 committed:
Add BOM to InitiativeGeoData and response example files

Added a UTF-8 Byte Order Mark (BOM) to the InitiativeGeoData.cs and InitiativeGeoDataResponseExample.cs files to ensure proper encoding handling.

cgalvist commented on src/Application/DTOs/Initiatives/InitiativeCoordinatesDto.cs:
```diff
@@ -0,0 +1,26 @@
+    /// <summary>
+    /// Initiative coordinates [longitude, latitude].
+    /// </summary>
+    public double[] Coordinate { get; set; } = Array.Empty<double>();
```
La inicialización se puede simplificar con el valor `[]`

cgalvist commented on src/Core/Domain/Entities/Initiatives/InitiativeGeoData.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// Initiative coordinates [longitude, latitude].
+    /// </summary>
+    public double[] Coordinate { get; set; } = Array.Empty<double>();
```
La inicialización se puede simplificar con el valor `[]`

cgalvist commented on src/Application/Services/Statistics/GeneralStatisticsService.cs:
Veo que esta línea y el comentario `Convert from km² to hectares` se repiten 3 veces en esta función.
Recomiendo crear una función estática que tenga ese comentario y que permita convertir de kilómetros cuadrados a hectáreas.
La función puede estar en alguna clase de la carpeta `Utils`

cgalvist reviewed (changes requested):
Hola.
A nivel funcional considero que está bien, pero dejo algunos comentarios adicionales de cosas que se podrían mejorar.

denarch22 committed:
Refactor initiative coordinates DTO usage and area conversion

Removed InitiativeCoordinatesDto and replaced its usage with InitiativeGeoData throughout the codebase. Updated GeneralStatisticsService to use a new utility method in GeometryUtils for converting square kilometers to hectares. Adjusted related example and entity code to align with these changes.

denarch22 committed:
Merge branch 'develop' into darmenta/LIB-352

cgalvist reviewed (approved):
Ya aplicaste las sugerencias. Muchas gracias 🙂
